lib/util/pre_receive_util.js

```diff
@@ -0,0 +1,227 @@
+        }
+    }
+
+    getSyntheticBranchForCommit(commit) {
+        return "refs/git-meta/synthetic/meta-" + commit.id();
+    }
+
+    *checkSubmodule(repo, commit, submoduleEntry) {
+        const submodulePath = submoduleEntry.path();
+        const submodule = yield NodeGit.Submodule.lookup(repo, submodulePath);
+        const url = submodule.url();
+        const localPath = yield *this.urlToLocalPath(repo, url);
+        const submoduleRepo = yield NodeGit.Repository.open(localPath);
+        const branch = this.getSyntheticBranchForCommit(commit);
+        try {
+            const subrepoCommit =
+                yield submoduleRepo.getReferenceCommit(branch);
+            return subrepoCommit.id().equal(submoduleEntry.id());
+        } catch (e) {
+            return false;
+        }
+    }
+
+    *checkTree(repo, commit, parentTrees, tree) {
+        const check = this;
+        return yield async.every(tree.entries(), function *(entry) {
+            if (entry.isFile()) {
+                return true;
+            }
+
+            const parentEntries = flatmap(parentTrees, function (parentTree) {
+                try {
+                    return parentTree.getEntryByName(entry.name());
+                } catch (e) {
+                    return null;
+                }
+            });
+
+            function isSameAsTree(parentEntry) {
+                try {
+                    const myEntry = tree.entryByName(entry.name());
+                    return myEntry.id() === parentEntry.id() ;
+                } catch (e) {
+                    return false;
+                }
+            }
+
+            if (!parentEntries.some(isSameAsTree)) {
+                /* none of the parent entries match this entry */
+                if (entry.isTree()) {
+                    const entryParentTrees = parentEntries.filter(
+                        function isTree(entry) {
+                            return entry.isTree();
+                        });
+                    const tree = yield entry.getTree();
+                    return yield *check.checkTree(repo, commit,
+                                                  entryParentTrees, tree);
+                } else {
+                    return yield *check.checkSubmodule(repo, commit, entry);
+                }
+            }
+        });
+    }
+
+    /**
+     * Recurse over the parents of a commit, checking each for git-meta's
+     * invariants. Returns when all (transitive) parents have already
+     * been marked as checked, or a failure is found.
+     *
+     * Returns true if the check passes on all branches; false if any
+     * fail.
+     */
+    *parentLoop(repo, commit, oldSha, handled) {
+        if (commit.id() in handled)  {
+            return true;
+        }
+        handled[commit.id()] = 1;
+        if (oldSha === commit.id().toString()) {
+            return true;
+        }
+
+        const ok = yield GitUtil.readNote(repo, NOTES_REF, commit.id());
```

Comment:
Is it worth it to check that the note says "ok"? Could the note ever say something else? Should we even be writing "ok" to the note instead of an empty string?